pkg/cloudprovider/provider/vultr/provider.go

```diff
@@ -155,64 +245,188 @@ func (p *provider) Validate(ctx context.Context, _ *zap.SugaredLogger, spec clus
+	}
+	resp.Body.Close()
+	for _, instance := range instances {
+		if sets.NewString(instance.Tags...).Has(string(machine.UID)) {
+			return &vultrPhysicalMachine{instance: &instance}, nil
 		}
 	}
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
 
+func (p *provider) getVirtualMachine(ctx context.Context, c *Config, machine *clusterv1alpha1.Machine) (*vultrVirtualMachine, error) {
 	client := getClient(ctx, c.APIKey)
 
-	instances, _, err := client.Instance.List(ctx, &govultr.ListOptions{
+	instances, _, resp, err := client.Instance.List(ctx, &govultr.ListOptions{
 		Tag: string(machine.UID),
 	})
 	if err != nil {
-		return nil, vltErrorToTerminalError(err, "failed to list servers")
+		return nil, vltErrorToTerminalError(resp.StatusCode, err)
 	}
+	resp.Body.Close()
 
 	for _, instance := range instances {
-		for _, tag := range instance.Tags {
-			if tag == string(machine.UID) {
-				return &vultrInstance{instance: &instance}, nil
-			}
+		if sets.NewString(instance.Tags...).Has(string(machine.UID)) &&
+			instance.Label == machine.Name {
+			return &vultrVirtualMachine{instance: &instance}, nil
 		}
 	}
 
 	return nil, cloudprovidererrors.ErrInstanceNotFound
 }
 
 func (p *provider) Get(ctx context.Context, _ *zap.SugaredLogger, machine *clusterv1alpha1.Machine, _ *cloudprovidertypes.ProviderData) (instance.Instance, error) {
-	return p.get(ctx, machine)
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+	if !c.PhysicalMachine {
+		return p.getVirtualMachine(ctx, c, machine)
+	}
+
+	return p.getPhysicalMachine(ctx, c, machine)
 }
 
 func (p *provider) GetCloudConfig(_ clusterv1alpha1.MachineSpec) (config string, name string, err error) {
 	return "", "", nil
 }
 
+func (p *provider) waitForInstanceCreation(ctx context.Context, c *Config, instance instance.Instance, machine *clusterv1alpha1.Machine) error {
+	return wait.PollUntilContextTimeout(ctx, createCheckPeriod, createCheckTimeout, false, func(ctx context.Context) (bool, error) {
+		var err error
+		if !c.PhysicalMachine {
+			_, err = p.getVirtualMachine(ctx, c, machine)
+		} else {
+			_, err = p.getPhysicalMachine(ctx, c, machine)
+		}
+
+		if err != nil && err != cloudprovidererrors.ErrInstanceNotFound {
+			if cloudprovidererrors.IsNotFound(err) {
+				// Continue the loop as the instances was succesfully fetched
+				// just that our instance was not found
+				return false, nil
+			}
+			if isTerminalErr, _, _ := cloudprovidererrors.IsTerminalError(err); isTerminalErr {
+				return true, err
+			}
+			// Wait for some time as instance creation is successful
+			// just that we are not able to fetch it
+			time.Sleep(createCheckFailedWaitPeriod)
+			return false, fmt.Errorf("instance %q created but controller failed to fetch instance details", instance.Name())
```

Comment:
I think if you return an error, the poll loop stops and returns the error - I don't think that's the intention here; it might be better to log the error and return nil instead so it keeps retrying.

Reply:
As this error is returned when there is a problem with the vultr client itself and the List() request fails inside getPhysicalMachine() or getVirtualMachine().
I feel that that kind of error should not be ignored and loop should be broken returning the reason as to why the client returned an error. It could happen that if vultr server is down or the client is misconfigured, this wait function might never return. What do you think?

Reply:
Yeah, I suppose it could go either way. There is still a possibility that Vultr is temporarily unreachabler due to some API issue (gateway failure, rate limiting, scheduled maintenance) and we could still get into a duplicate instance scenario.

I think the wait function still has a hard timeout of 5 minutes, right? So I think it'd probably eventually bail out either way. Maybe 5 minutes is a bit excessive in this scenario; but at the same time, if the provider gets this far, then the instance create call completed without error, so at least in theory things are most likely out of sync if the code bails out.

I'm less worried about this particular flow, however, as it's definitely an edge case.

I saw some of the other providers go into a "Cleanup" cycle if they fail creation - is that something that would be a better option here? It might leave an orphaned instance in Vultr, but it wouldn't result in a "two instance" reconciliation problem as the machine for that label would be gone, right?

Reply:
Cleanup cycle when the Instance creation itself fails or the waiting for label fails?
In any case, what kind of metadata will we give the vultr API to delete instance? 
Tags might not have applied to the instance and can't give the label as the controller might have created another instance for the machine.

Reply:
An ideal way i could think of is to know whether machine-controller controlled label exists on the machine or not. Currently we can't know whether the UID label is present on machine or not. If we could know that, we can add "wait"s wherever (instance.label == machine.name && UID label DNE) either when creating instance or when getting the instance. If we can't, we(provider) gives error.
As for the cleanup i feel that, maybe we can do it syncronously after the "wait"s and delete the instance if the wait ended with no error and no success. But that would need controller changes.

I think the cleanup could be done here as well, after the wait, let me try that.

Reply:
Pushed.